### PR TITLE
Development idempotency header

### LINKED ISSUE #145: Support for Idempotent Requests
I need the ability to make Idempotent Requests as explained here:
https://stripe.com/docs/api/idempotent_requests?lang=curl

The problem is that these requests are a) possible whenever posting regardless of API and b) are put in the header of the request not the data section.

I think the way to implement is to give all "put" requests (create in our current paradigm) an optional Idempotent string value.  Unfortunately (or fortunately, depending on who you ask) Rust doesn't support default parameters, so while it will be optional, it will break the top level API.  Not much we can do about that though.

I'm working on the implementation now, let me know if you have any objects/comments when you're available.

### PR DESCRIPTION
**Pull request by erichCompSci**
Closes #145

Okay, Arlyon

Here is the simplified idempotency API.  It's much simpler and I can tell you that it works, I've tested myself.

It's so small, I didn't bother making it a feature and it doesn't break backward compatibility, so I think that's good.

### CODE REVIEW

File: src/client/async_std.rs [hunk 1]
```diff
@@ -67,6 +67,15 @@ impl Client {
         client
     }
 
+    /// Clones a new client with idempotency key headers.
+    ///
+    /// For short living idempotency keys, they only matter on push requests
+    pub fn with_idempotency_key(&self, idempotency_key: String) -> Client {
+        let mut client = self.clone();
```
Comment (arlyon) [inline src/client/async_std.rs:74]: I would suggest allowing the user to make the decision to clone, and instead move `self`. That way the user can re-use the same object. Something along the lines of this (note we take self, not a borrow):

```rust
pub fn with_idempotency_key(self, ) -> Self {
  Self {
    // update params with key
    ..self
  }
}

Comment (arlyon) [reply src/client/async_std.rs:74]: Then in the user facing api it looks like this:

```rust
let client = ...
let idempotent_client = client.clone().with_idempotency_key("test".to_string());
// here both clients are valid
// alternatively
let idempotent_client = client.with_idempotency_key("test".to_string());
// now we 'move' the original client

File: src/client/blocking.rs [hunk 2]
```diff
@@ -55,6 +55,15 @@ impl Client {
         Client { inner: self.inner.with_headers(headers), runtime: self.runtime.clone() }
     }
 
+    /// Clones a new client with idempotency key headers.
+    ///
+    /// For short living idempotency keys, they only matter on push requests
+    pub fn with_idempotency_key(&self, idempotency_key: String) -> Client {
+        let mut client = self.clone();
```
Comment (arlyon) [inline src/client/blocking.rs:62]: Similarly here, we are cloning the provided client, then calling inner.with_idempotency_key which performs an additional clone. Allowing the user to choose would avoid this case.

**Review by arlyon — CHANGES_REQUESTED** [review]
Great stuff! Minor changes and I'll get this merged. Afterwards I am going to impl retries on top of this.

**Comment by erichCompSci** [comment]
Hey Arlyon, 

I had to add a error handling commit to move forward on some other stuff I'm trying to do.  I'm still waiting for your response to my previous comment, please ignore this pull request until you respond.

File: src/client/blocking.rs (hunk 2 above)
Comment (arlyon) [reply src/client/blocking.rs:62]: Hi! Decided to do some reading to back up my intuition: a lot of rust libraries that provide APIs with `with_` will move the object. Examples:

- https://cs.github.com/hyperium/hyper/blob/055b4e7ea6bd22859c20d60776b0c8f20d27498e/src/server/conn.rs?q=with_#L905
 - https://cs.github.com/hyperium/hyper/blob/055b4e7ea6bd22859c20d60776b0c8f20d27498e/src/server/server.rs?q=with_#L165
 - https://cs.github.com/tokio-rs/tracing/blob/5f806286bdef217fefa0e6c3f47a6267f0855e2b/tracing-tower/src/lib.rs?q=with_#L80
 - https://cs.github.com/tokio-rs/tracing/blob/5f806286bdef217fefa0e6c3f47a6267f0855e2b/tracing/src/instrument.rs?q=with_#L186

Thanks for bringing up the inconsistency. I think we can, in deciding to follow that convention, also update the code in places that until now haven't been.

Comment (erichCompSci) [reply src/client/blocking.rs:62]: Okay, great, I'll fix both up and resubmit the pull request when I get off from work later tonight.

**Pull request closed without merging**